File: sdk/samples/WCFSamples/TechnologySamples/Scenario/Router/CPP/service-echo/EchoService.cpp
// ----------------------------------------------------------------------------
// Copyright (C) 2003-2005 Microsoft Corporation, All rights reserved.
// ----------------------------------------------------------------------------

using namespace System;
using namespace System::ServiceModel;
using namespace System::ServiceModel::Dispatcher;
using namespace System::Diagnostics;
using namespace System::Configuration;

namespace Microsoft{
namespace ServiceModel{
namespace Samples{

    [System::ServiceModel::ServiceContract(Namespace=L"http://Microsoft.ServiceModel.Samples/")]
    public interface class IEchoService
    {
        [System::ServiceModel::OperationContract(IsOneWay=false)]
        System::String ^ Echo(System::String ^ message);
    };




    [System::ServiceModel::ServiceBehavior(InstanceContextMode=System::ServiceModel::InstanceContextMode::Single)]
    public ref class EchoService : public IEchoService
    {

    public: 
        EchoService()
        {
        }

        virtual System::String ^ Echo(System::String ^ message)
        {
            return message;
        }
    };
}
}
}



int main(array<System::String ^> ^args)
{	
    ServiceHost ^ serviceHost = gcnew ServiceHost( Microsoft::ServiceModel::Samples::EchoService::typeid );

    serviceHost->Open();


    for (int i = 0; i < serviceHost->ChannelDispatchers->Count; i++)
    {
        ChannelDispatcher ^ channelDispatcher = safe_cast<ChannelDispatcher^>( serviceHost->ChannelDispatchers[ i ] );

        for ( int j = 0; j < channelDispatcher->Endpoints->Count; j++ )
        {
            EndpointDispatcher ^ endpointDispatcher = channelDispatcher->Endpoints[ j ];
            Console::WriteLine( L"Listening on {0}...", endpointDispatcher->EndpointAddress->ToString() );
        }        
    }


    Console::WriteLine();
    Console::WriteLine(L"Press Enter to exit...");
    Console::ReadLine();
    serviceHost->Close();

    return 0;
}

File: sdk/samples/WCFSamples/TechnologySamples/Extensibility/Rules/WCF_Router/cs/service-echo/EchoService.cs
//  Copyright (c) Microsoft Corporation. All rights reserved.

using System;
using System.ServiceModel;
using System.ServiceModel.Dispatcher;

namespace Microsoft.ServiceModel.Samples
{
    [ServiceContract(Namespace = "http://Microsoft.ServiceModel.Samples/")]
    public interface IEchoService
    {
        [OperationContract(IsOneWay = false)]
        string Echo(string message);
    }

    [ServiceBehavior(InstanceContextMode = InstanceContextMode.Single)]
    public class EchoService : IEchoService
    {
        string IEchoService.Echo(string message)
        {
            return message;
        }
    }

    public class ServiceDriver
    {
        public static void Main(string[] args)
        {
            ServiceHost serviceHost = new ServiceHost(typeof(EchoService));
            serviceHost.Open();

            for (int i = 0; i < serviceHost.ChannelDispatchers.Count; i++)
            {
                ChannelDispatcher channelDispatcher = serviceHost.ChannelDispatchers[i] as ChannelDispatcher;
                if (channelDispatcher != null)
                {
                    for (int j = 0; j < channelDispatcher.Endpoints.Count; j++)
                    {
                        EndpointDispatcher endpointDispatcher = channelDispatcher.Endpoints[j];
                        Console.WriteLine("Listening on " + endpointDispatcher.EndpointAddress + "...");
                    }
                }
            }

            Console.WriteLine();
            Console.WriteLine("Press Enter to exit...");
            Console.ReadLine();
        }
    }
}
